US-11: Search Start and End Locations

osahtout opened the pull request:
This request implements US-11 (#18).

A few new components were added for full functionality of the search component: search bar component, autocomplete component, omnibox component and the start travel component.

r-saba committed:
Add Search Component #18

r-saba committed:
Add autocomplete component #18

r-saba committed:
Finish search bar user interface #18

r-saba committed:
Add omnibox-directions component #18

osahtout committed:
Add search query for destination markers #18

Save destination location #18

osahtout committed:
Add time picker component #18

osahtout committed:
Run pre-commit and refactored #18

refactor searching function #18

r-saba committed:
change omnibox-direction logic to allow destination searches #18

r-saba committed:
Add tests for search component #18

r-saba committed:
Add selectable markers for start/end location #18

Fix current location and travel plan bugs #18

r-saba committed:
Add tests for all search components #18

Add more tests and change time picker #18

osahtout committed:
Refactor and fix pre-commit #18

Run lint and run pre-commit #18

osahtout committed:
Fix android time picker bug #18

fix snapshot bug #18

cristian-aldea commented on package.json:
```diff
@@ -11,7 +11,11 @@
+  "eslintIgnore": [
+    "node_modules/*"
+  ],
   "dependencies": {
```
I see that some of these dependencies aren't being used. make sure to remove them if that is the case

cristian-aldea commented on src/components/map-overlays/custom-marker.component.tsx:
```diff
@@ -8,9 +8,10 @@ interface IProps {
+  markerType: "campus" | "building" | "poi" | "destination";
 }
 let backgroundColor: string = CONCORDIA_RED;
+let icon: ImageURISource;
```
This icon doesn't seem to be used anywhere.

cristian-aldea commented on src/components/search/autocomplete.component.tsx:
```diff
@@ -0,0 +1,85 @@
+  container: {
+    position: "absolute",
+    flex: 1,
+    width: Dimensions.get("window").width - 30,
```
There are constants available for the screen width in constants/style.ts

bit172 commented on src/components/search/search.component.tsx:
```diff
@@ -0,0 +1,129 @@
+              style={styles.lineSeperator}
+              source={require("../../../assets/line-separator.png")}
+            />
+            <Image source={require("../../../assets/search.png")} />
```
These icons should be SVGs, not images. Images appear pixelated on larger screens.

cristian-aldea replied:
Good point, +1 from me.

Also, don't forget that you have access to an enormous library of icons from expo
https://docs.expo.io/versions/latest/guides/icons/

cristian-aldea commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -0,0 +1,323 @@
+import React, { useEffect } from "react";
```
This file is huge and the name OmniboxDirections is really confusing

cristian-aldea commented on src/components/search/autocomplete.component.tsx:
```diff
@@ -0,0 +1,85 @@
+import React from "react";
```
This <Autocomplete> component is used three times shared between omnibox-directions.component.tsx and search.component.tsx. I wonder if that could be simplified

cristian-aldea commented on src/components/search/search.component.tsx:
```diff
@@ -0,0 +1,129 @@
+/**
+ * Dynamic height adjustment of parent. Without this, autoComplete will not be pressable
+ */
+export const getAutoCompleteHeight = (autoCompleteValues?: string[]) => {
```
This method is duplicated from omnibox-directions component to handle the auto-complete. find a way to avoid this duplication if possible

cristian-aldea replied:
Actual, please also move this to a service along with the other method of the same name

TheGreatMarkus commented on src/components/search/start-travel.component.tsx:
```diff
@@ -0,0 +1,59 @@
+    <View style={styles.container}>
+      <TouchableOpacity
+        onPress={() => setStartTravelPlan(true)}
+        style={{ width: 80, height: 80 }}
```
No styling in the jsx please. Also the button is too close to the right side and too big (on android at least).

cristian-aldea commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -49,6 +53,14 @@ const MapScreen = () => {
+  const [initialLocation, setInitialLocation] = useState<POI | UserLocation>(
+    null
+  );
+  const [markerSetsDestination, setMarkerSetsDestination] = useState<Boolean>(
```
what does markerSetsDestination do exactly? consider leaving a comment.

osahtout replied:
this is an example of bad naming that should be changed

cristian-aldea commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -189,11 +201,67 @@ const MapScreen = () => {
   useEffect(() => {
     setCurrentRegion(getCampusById(CampusId.SGW).region);
+    if (!currentLocation) {
+      getCurrentLocationAsync().then(location => {
```
I don't think we need the user's location at all times. Consider only calling this method when you actually need the location (like when starting a travel plan)

cristian-aldea replied:
Also, please add a catch() for every call to getCurrentLocationAsync

cristian-aldea commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -189,11 +201,67 @@ const MapScreen = () => {
   }, []);
 
+  /**
+   *  Set the value of destination or initial locaiton depending
```
typos

RGPosadas commented on .eslintrc.js:
```diff
@@ -1,4 +1,5 @@
 module.exports = {
+  ignorePatterns: ["node_modules/*"],
```
So which one works finally, this, or .eslintignore, or both?

RGPosadas commented on .eslintrc.js:
```diff
@@ -1,4 +1,5 @@
 module.exports = {
+  ignorePatterns: ["node_modules/*"],
```
So which one works finally? This, `.eslintignore`, or in `package.json`?
Please pick only one

osahtout replied:
I'm not sure because I removed all of them and it still works

cristian-aldea commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -189,11 +201,67 @@ const MapScreen = () => {
+   * Filters an array of POIs based on users input
+   * @param userInput
+   */
+  const queryText = (userInput, setAutocomplete, onChangeText) => {
```
This shouldn't be in the in the mapscreen imo. It could be made pure and placed in a service

cristian-aldea commented on src/types/main.ts:
```diff
@@ -5,6 +5,12 @@ export interface Location {
   longitude: number;
 }
 
+export interface UserLocation {
```
Is this new interface really necessary? I'm not 100%, but I feel that this info could be represented with already existing types (maybe just a location?)

cristian-aldea reviewed (changes requested):
Overall pretty cool, but still needs a lot of work. I'm getting a bunch of bugs and exception need to be handled correctly.

RGPosadas commented on __tests__/autocomplete.test.js:
```diff
@@ -0,0 +1,55 @@
+import renderer from "react-test-renderer";
+import { shallow } from "enzyme";
+
+let mockPOIs = [
```
BLOCKING:
No need to mock constant data. You can compare them directly. Please change to use the POI data file, so it's easier to see if tests failed due to different POI structure.
See example here: https://github.com/RGPosadas/WayFinder/blob/TASK-10/setup-detox/__tests__/custom-marker.test.js#L54

cristian-aldea replied:
Other solution would be to convert tests to typescript. might just be this file for now or could be done in a seperate issue.

r-saba replied:
Converted this test to typescript

RGPosadas commented on __tests__/omnibox-directions.test.js:
```diff
@@ -0,0 +1,231 @@
+import { shallow } from "enzyme";
+import { Platform } from "react-native";
+
+let mockPOIs = [
```
BLOCKING
Same comment
See: https://github.com/RGPosadas/WayFinder/blob/TASK-10/setup-detox/__tests__/custom-marker.test.js#L54

r-saba replied:
test converted to typescript

RGPosadas commented:
### Functionality
* BLOCKING 1: Location button should disappear once the search mode is on. (Or should location button stay and we move the start navigation button elsewhere? Let's discuss)
* BLOCKING 2: If user denies location services, I get this error:
```
[Unhandled promise rejection: Error: Location Permission Denied]
* http://packager.7k-kx9.anonymous.wayfinder.exp.direct/node_modules/expo/AppEntry.bundle?platform=android&dev=true&minify=false&hot=false:155064:28 in getCurrentLocationAsync$
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:274:30 in invoke
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:135:28 in invoke
- node_modules/regenerator-runtime/runtime.js:145:19 in PromiseImpl.resolve.then$argument_0
- node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:25 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:146:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:194:17 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:458:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:407:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:384:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:142:17 in __guard$argument_0
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue
```
* BLOCKING 3: Once I choose a time, it becomes `DEPART AT: AN:AN`. If I click on `DEPART AT: AN:AN` again, I get this error on Google Pixel 3:
```
[Unhandled promise rejection: Error: Exception in HostFunction: Malformed calls from JS: field sizes are different.]
* [native code]:null in <unknown>
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:337:39 in enqueueNativeCall
- node_modules/react-native/Libraries/BatchedBridge/NativeModules.js:137:10 in fn
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:487:27 in createTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:253:16 in setTimeout
- node_modules/promise/setimmediate/rejection-tracking.js:47:10 in Promise._87
- node_modules/promise/setimmediate/core.js:169:16 in reject
- node_modules/promise/setimmediate/core.js:211:11 in doResolve
- node_modules/promise/setimmediate/core.js:66:12 in Promise
- node_modules/react-native/Libraries/BatchedBridge/NativeModules.js:101:25 in Promise$argument_0
- node_modules/@react-native-community/datetimepicker/src/timepicker.android.js:37:4 in TimePickerAndroid.open
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:274:30 in invoke
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:135:28 in invoke
- node_modules/regenerator-runtime/runtime.js:170:17 in PromiseImpl$argument_0
- node_modules/promise/setimmediate/core.js:45:7 in tryCallTwo
- node_modules/promise/setimmediate/core.js:200:23 in doResolve
- node_modules/promise/setimmediate/core.js:66:12 in Promise
- node_modules/regenerator-runtime/runtime.js:169:31 in PromiseImpl$argument_0
- node_modules/regenerator-runtime/runtime.js:192:38 in enqueue
- node_modules/regenerator-runtime/runtime.js:219:8 in exports.async
* http://packager.7k-kx9.anonymous.wayfinder.exp.direct/node_modules/expo/AppEntry.bundle?platform=android&dev=true&minify=false&hot=false:147315:42 in open
- node_modules/@react-native-community/datetimepicker/src/datetimepicker.android.js:34:4 in RNDateTimePicker
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:10696:27 in renderWithHooks
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:13481:6 in mountIndeterminateComponent
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:20459:25 in beginWork$$1
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:19370:24 in performUnitOfWork
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:19347:39 in workLoopSync
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:18997:22 in renderRoot
* [native code]:null in renderRoot
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:18709:28 in runRootCallback
* [native code]:null in runRootCallback
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:5642:32 in runWithPriority$argument_1
- node_modules/scheduler/cjs/scheduler.development.js:643:23 in unstable_runWithPriority
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:5638:22 in flushSyncCallbackQueueImpl
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:5627:28 in flushSyncCallbackQueue
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:18796:28 in batchedUpdates$1
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:2709:30 in batchedUpdates
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:2794:17 in batchedUpdates$argument_0
- node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js:2870:28 in receiveTouches
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:436:47 in __callFunction
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:111:26 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:384:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:110:17 in __guard$argument_0
* [native code]:null in callFunctionReturnFlushedQueue
```
![image](https://user-images.githubusercontent.com/33991121/77478719-d880f800-6df4-11ea-90b0-d06fed2735e3.png)


### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All tests pass
Worried about this test output error. Have you looked into it?
![image](https://user-images.githubusercontent.com/33991121/77478340-32cd8900-6df4-11ea-99ac-69056729801b.png)
- [] Implementation matches the corresponding mockup from Figma (see BLOCKING 1)

### ACCEPTANCE CRITERIA
- [x] Indoor locations should be available on the map to be searched and selected
- [x] The first location that the user clicks or searches for will become the destination point.
- [x] If location services are available, selecting the user's current location as the starting point is done automatically.
- [] If the starting point isn't set, the first location that the user clicks or searches for will become the starting point.
- [] After the starting point is set, markers indicating the starting and destination location will be shown on the map.

### USER FLOW ACCEPTANCE TEST
- [x] Tap on search bar
- [x] User can search or tap on a destination room
- [x] If location services are on, starting location is filled in with user's current location
- [] If location services are not on, starting location is filled in by user (by search or tap) (see BLOCKING 2)

cristian-aldea commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -0,0 +1,323 @@
+  },
+  picker: {
+    marginLeft: Platform.OS === "ios" ? 38 : 36,
+    borderColor: "#DEDEDE",
```
consider moving this to a constant

cristian-aldea commented on src/components/map-overlays/map-overlays.component.tsx:
```diff
@@ -108,7 +122,7 @@ const MapOverlays = ({
       {zoomLevel === ZoomLevel.INDOOR ? (
```
Add condition where you only show this if no travel plans are set

cristian-aldea commented on src/components/map-overlays/map-overlays.component.tsx:
```diff
@@ -124,15 +138,39 @@ const MapOverlays = ({
+                );
+              }
+              return null;
+            })}
```
Add another blow which display those two marker only if there is currently a travel plan

osahtout committed:
Addressed many pull request comments #18

fix pre-commit #18

addressed more comments #18

Resolve pr comments #18

clean up #18

bit172 commented on src/components/map-overlays/custom-marker.component.tsx:
```diff
@@ -1,5 +1,5 @@
 import React from "react";
-import { StyleSheet, Text, View } from "react-native";
+import { StyleSheet, Text, View, ImageURISource } from "react-native";
```
ImageURISource is unused.

bit172 commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -0,0 +1,288 @@
+            {showTimePicker && (
+              <DateTimePicker
+                testID="dateTimePicker"
+                value={date || initDate}
```
Can this value be set in the useEffect instead of checking?

bit172 commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -0,0 +1,288 @@
+      });
+    }
+  });
+
```
![image](https://user-images.githubusercontent.com/33991121/77478340-32cd8900-6df4-11ea-99ac-69056729801b.png) I suspect this is caused by the ordering of react hooks, but the fixes I tried did not work. More on this here: https://reactjs.org/docs/hooks-rules.html Not sure if this will cause problems in the future.

r-saba replied:
This was the only way I found to test useState hook and the fixes we've tried have not worked either. This increases our test coverage and should problems arise from this test in the future we can remove it.

bit172 commented on src/components/search/search.component.tsx:
```diff
@@ -0,0 +1,117 @@
+        <View style={styles.view}>
+          <View style={styles.view}>
+            <SimpleLineIcons name="menu" color="#AA2B45" size={26} />
+            <SeparatorSVG style={styles.lineSeperator} />
```
Thanks for the SVGs! 😅

r-saba replied:
So for omnibox-directions.component I've kept the PNGs but I used higher res images and scaled them down to keep the image quality. Had to do this because snapshots were failing with svg's for that component only and I was unable to find a fix at the moment.

r-saba committed:
Refactor marker logic #18

Co-authored-by: Omar <osahtout@hotmail.com>
Co-authored-by: Cristian Aldea <aldea.cristian@outlook.com>

r-saba committed:
add missing icons to omnibox-directions #18

Addressed many more pull request comments #18

fix pre-commit #18

addressed more comments #18

Resolve pr comments #18

clean up #18

address pr comments #18

TheGreatMarkus commented on declarations.d.ts:
```diff
@@ -0,0 +1,6 @@
+declare module "*.svg" {
```
What was the reason behind including this file?

cristian-aldea commented on declarations.d.ts:
```diff
@@ -0,0 +1,6 @@
+declare module "*.svg" {
```
What was the reason behind including this file?

r-saba replied:
Since react native does not support SVG's we are using a library which allows us to import SVG's like in react. This is needed for typescript to recognize the SVG's properly. As per the library documentation here https://github.com/kristerkari/react-native-svg-transformer#using-typescript

cristian-aldea commented on src/components/map-overlays/custom-marker.component.tsx:
```diff
@@ -8,7 +8,7 @@ interface IProps {
   onPress: () => void;
   text: string;
-  markerType: "campus" | "building" | "poi";
+  markerType: "campus" | "building" | "poi" | "destination";
```
Is this additional type still needed?

r-saba replied:
@osahtout This is related to the markerType comment below. Can you remove this if not needed

cristian-aldea commented on src/components/map-overlays/map-overlays.component.tsx:
```diff
@@ -130,7 +141,35 @@ const MapOverlays = ({
+        <>
+          {getAllPOI()
+            .filter(marker => {
+              if (startTravelPlan) {
```
This extra check for startTravelPlan is not necessary

osahtout replied:
it is since we are rendering the markers at any zoom level. it would also not be recommended to render the marker twice

cristian-aldea replied:
But you already do this check on line 152, no? This block is only ever activated when startTravelPlan is true. Why do a second check on line 156?

osahtout replied:
ahh right, I didn't see it in github comments

cristian-aldea commented on src/components/map-overlays/map-overlays.component.tsx:
```diff
@@ -130,7 +141,35 @@ const MapOverlays = ({
+            })
+            .map(poi => (
+              <CustomMarker
+                markerType="poi"
```
marker type should be "destination", no?

cristian-aldea commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -0,0 +1,297 @@
+} from "../../constants/style";
+import { getOmniboxAutoCompleteHeight } from "../../services/autoCompleteHeight.service";
+
+const getTimeToString = (date: Date) => {
```
This function is pure and could be ported to an outside service file

cristian-aldea commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -0,0 +1,297 @@
+        </View>
+        {((autoCompleteValues && value !== "") ||
+          (autoCompleteValuesDest && destinationValue !== "")) && (
+          <Autocomplete
```
Thanks for simplifying this :)

cristian-aldea commented on src/services/autoCompleteHeight.service.ts:
```diff
@@ -0,0 +1,47 @@
+import { Platform } from "react-native";
```
please follow the file name convention we've had up until now. Rename the file to: auto-complete-height.service.ts.

cristian-aldea commented on src/services/queryUserIntput.service.ts:
```diff
@@ -0,0 +1,21 @@
+import { POI } from "../types/main";
```
please follow the file name convention we've had up until now. Rename the file to: query-user-input.service.ts.

cristian-aldea reviewed (commented):
Much better! Left some more suggestions

RGPosadas reviewed (changes requested):
* Not blocking: BLOCKING 2 from my previous review is now fixed, however the same error appears when you press the location button since it isn't handled
May or may not be your scope, but if it's easily implementable I wouldn't mind having it in this PR 

* BLOCKING: I still get `DEPART AT: AN:AN` instead of the time that I manually put in (Pixel 3)

TheGreatMarkus commented:
Not a review, but I just thought of a way to make the marker logic for map-overlays more simple. Gonna implement it real quick

cristian-aldea commented on src/components/search/autocomplete.component.tsx:
```diff
@@ -0,0 +1,91 @@
+  autoCompleteValues,
+  style,
+  setLocation,
+  testID
```
testID isn't being used. Consider removing it.

cristian-aldea commented on src/components/search/autocomplete.component.tsx:
```diff
@@ -0,0 +1,91 @@
+  testID
+}: IProps) => {
+  return (
+    <View style={[styles.container, { ...style }]}>
```
From what I know, using StyleSheet.flatten([...]) is best practice.

cristian-aldea commented on __tests__/omnibox-directions.test.tsx:
```diff
@@ -0,0 +1,274 @@
+  });
+
+  it("should run provided function when back arrow pressed", () => {
+    const mockSetDestination = jest.fn();
```
Rename the variables for added clarity:
* mockSetDestination => setEndLocation
* mocSetInitialLocation [sic.] => mockSetStartLocation
* mockSetIsDestinationFocused => mockSetEndLocationFocused

Essentially, make sure the mock is called `mock[nameOfOriginalFunction]`

r-saba replied:
They were but I think the function names were changed in the previous commit? Will update

cristian-aldea commented on __tests__/omnibox-directions.test.tsx:
```diff
@@ -0,0 +1,274 @@
+    expect(mockQueryText.mock.calls[0][0]).toBe("H805");
+  });
+
+  it("should call the provided functin when initial location textInput is focused", () => {
```
Typo, also consider renaming this to: `should call the provided function on focus of startLocation TextInput`

cristian-aldea commented on __tests__/omnibox-directions.test.tsx:
```diff
@@ -0,0 +1,274 @@
+  });
+
+  it("should call the provided functin when initial location textInput is focused", () => {
+    const mockSetIsDestinationFocused = jest.fn();
```
As before, rename mock functions to `mock[nameOfOriginalFunction]`

cristian-aldea commented on __tests__/omnibox-directions.test.tsx:
```diff
@@ -0,0 +1,274 @@
+    expect(mockQueryText.mock.calls[0][0]).toBe("H805");
+  });
+
+  it("should call the provided functin when endLocation location textInput is focused", () => {
```
Typo, also consider renaming this to: `should call the provided function on focus of endLocation TextInput`

cristian-aldea commented on __tests__/omnibox-directions.test.tsx:
```diff
@@ -0,0 +1,274 @@
+  });
+
+  it("should call the provided functin when endLocation location textInput is focused", () => {
+    const mockSetIsDestinationFocused = jest.fn();
```
As before, rename mock functions to mock[nameOfOriginalFunction]

RGPosadas reviewed (changes requested):
### BLOCKING 4
* With location services off, it doesn't autocomplete the search when I click on a POI, and returns an unhandled promise error
* With location services on, clicking on a POI doesn't autocomplete
* This used to work in your earlier implementation, so this is a regression
* Tested on: Google Pixel 3 and iPhone 11 Max Pro

#### Steps to recreate the regression (with and without location services)
* Open the project
* Zoom in to show POIs
* Tap on the search bar, but do not write anything
* Tap on a POI
* With location services on, nothing happens (no autocomplete)
* With location services off, unhandled promise error occurs

#### Error log
```
[Unhandled promise rejection: Error: Location Permission Denied]
* http://packager.ej-cv2.ritchellegmp.wayfinder.exp.direct/node_modules/expo/AppEntry.bundle?platform=android&dev=true&minify=false&hot=false:175204:28 in getCurrentLocationAsync$
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:274:30 in invoke
- node_modules/regenerator-runtime/runtime.js:45:44 in tryCatch
- node_modules/regenerator-runtime/runtime.js:135:28 in invoke
- node_modules/regenerator-runtime/runtime.js:145:19 in PromiseImpl.resolve.then$argument_0
- node_modules/promise/setimmediate/core.js:37:14 in tryCallOne
- node_modules/promise/setimmediate/core.js:123:25 in setImmediate$argument_0
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:146:14 in _callTimer
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:194:17 in _callImmediatesPass
- node_modules/react-native/Libraries/Core/Timers/JSTimers.js:458:30 in callImmediates
* [native code]:null in callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:407:6 in __callImmediates
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:143:6 in __guard$argument_0
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:384:10 in __guard
- node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:142:17 in __guard$argument_0
* [native code]:null in flushedQueue
* [native code]:null in invokeCallbackAndReturnFlushedQueue
```

### Not Blocking
Not blocking, but could we position the search bar slightly lower? It interferes with notches (Google Pixel 3), excuse the JPEG
![image](https://user-images.githubusercontent.com/33991121/77805178-afac6d00-7057-11ea-9f2b-3b9b1f050913.png)

cristian-aldea commented on __tests__/omnibox-directions.test.tsx:
```diff
@@ -0,0 +1,274 @@
+      .onPress();
+
+    expect(mockSetshowTimePicker).toHaveBeenCalledTimes(1);
+  });
```
mockSetshowTimePicker has wrongcasing on line 183

We should avoid // @ts-ignore if possible.
Change line 187 to: `useStateSpy.mockImplementation(() => [null, mockSetshowTimePicker]);` The error will go away

Actually, this whole test is essentially testing the DateTimePicker component, which shouldn't be tested when testing the OmniboxDirections component. I suggest we remove it if possible.

osahtout replied:
This cannot be done since it needs an initial value to set and null can't be passed, explained here : https://dev.to/theactualgivens/testing-react-hook-state-changes-2oga

cristian-aldea commented on __tests__/omnibox-directions.test.tsx:
```diff
@@ -0,0 +1,274 @@
+    expect(mockSetshowTimePicker).toHaveBeenCalledTimes(1);
+  });
+
+  it("should get the height of the autoComplete component", () => {
```
This test should be moved to another file: `auto-complete-height.test.ts`

cristian-aldea commented on __tests__/search.test.js:
```diff
@@ -0,0 +1,32 @@
+    expect(mockQueryText.mock.calls[0][0]).toBe("H805");
+  });
+
+  it("should fucking do something", () => {
```
Ummmm, yeah. Rename this please

osahtout replied:
this is an example of bad naming

cristian-aldea commented on __tests__/search.test.js:
```diff
@@ -0,0 +1,32 @@
+    expect(height).toBe(135);
+  });
+
+  it("should do something else", () => {
```
Rename this please

cristian-aldea commented on __tests__/start-travel.test.js:
```diff
@@ -0,0 +1,23 @@
+    expect(tree).toMatchSnapshot();
+  });
+
+  it("should call a function on user input", () => {
```
Consider renaming this to `should call provided function on press`

osahtout replied:
Name doesn't make sense it's not an on press it's a change when the user starts typing (user input)

cristian-aldea replied:
I remembered wrong that this test had a simulated button press. that's my bad

cristian-aldea commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -191,9 +200,61 @@ const MapScreen = () => {
+        setEndLocation={setEndLocation}
+        setStartLocation={setStartLocation}
+        startLocation={startLocation}
+        queryText={queryText}
```
Why import queryText in MapScreen and then pass it to OmniboxDirections and Search? Just import it from each component.

r-saba replied:
Good point. A remnant of refactored code

r-saba replied:
After refactoring I was having trouble with getting the tests to work since this function involves state. At the moment I've kept it as is, if you're available to help with this let me know. Please leave a comment or resolve.

cristian-aldea commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -0,0 +1,301 @@
+   * @param event
+   * @param date
+   */
+  const onChange = (event, pickDate) => {
```
Please rename to onDateChange

cristian-aldea commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -191,9 +200,61 @@ const MapScreen = () => {
+    }
+  };
+
+  const setUserLocation = () => {
```
Add docs

cristian-aldea commented on src/services/auto-complete-height.service.ts:
```diff
@@ -0,0 +1,47 @@
+import { Platform } from "react-native";
+
+/**
+ * Dynamic height adjustment of parent. Without this, autocomplete will not be pressable
```
Add @params

cristian-aldea commented on src/services/auto-complete-height.service.ts:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Two function simply because all the inputs and values are different
+ */
+export const getSearchBarAutoCompleteHeight = (
```
Add @params

cristian-aldea reviewed (changes requested):
Alright, a lot of comments here, for cleanup mostly. Here are a few general fixes I'd like to see:

1. Move styling and colours to constants. Make sure there are no string literals as colours in components if possible, this will help us keep colouring consistent more easily and make refactoring the styling a lot easier.

2. Add docs to every single function and component you add. Just go through the [pr changes](https://github.com/RGPosadas/WayFinder/pull/62/files) and make sure each new function/component has some docs.

r-saba committed:
Add constants and address pr comments #18

Address pr comments and refactor code #18

fix bugs #18

r-saba committed:
fix current location display #18

r-saba committed:
Addressed final changes before merge #18

JimLau49 commented on __tests__/auto-complete-height.service.test.tsx:
```diff
@@ -0,0 +1,68 @@
+      (showTimePicker = null),
+      (endLocationInputValue = null)
+    );
+    expect(autoComplete).toBe(260);
```
What is this 260 corresponding to

r-saba replied:
`it("should get the correct height of the AutoComplete component", () => {
`
This test is making sure the height calculated is correct for the component. The height changes based on the number of search results being shown on the screen.

bit172 replied:
Thank your for the clarification

JimLau49 reviewed (dismissed):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All tests pass
- [x] Implementation matches the corresponding mockup from Figma

### ACCEPTANCE CRITERIA
* [x] Indoor locations should be available on the map to be searched and selected
* [x] The first location that the user clicks or searches for will become the destination point.
* [x] If location services are available, selecting the user's current location as the starting point is done automatically.
* [x] If the starting point isn't set, and the user places focus on the top text box, the next location that the user clicks or searches for will become the starting point.
* [x] After the starting and destination points are set, the user is allowed to start the travel plan
* [x] Once the user has started a travel plan, markers indicating the starting and destination location will be shown on the map.
* [x] Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=32%3A347)
* [x] Respect the [Definition of Done](https://github.com/RGPosadas/WayFinder/wiki/Definition-of-Done)


### USER FLOW ACCEPTANCE TEST
1. [x] Tap on the search bar
2. [x] User can search or tap on a point of interest to set it as the destination
3. [x] If location services are on, `starting location` is filled with the user's current location
    a. [x]  If location services are not on, `starting location` is filled in by the user. First tap on the top text box and then either search for a POI or tap on a POI on the map
4. [x] Once the start and end locations are set, press on the now visible start travel button on the bottom right
5. [x] Markers for the start and end location will be shown on the map

Left a question above, but otherwise it LGTM.
You guys really put a lot into this and it shows 🥇

cristian-aldea reviewed (dismissed):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All tests pass
- [x] Implementation matches the corresponding mockup from Figma

### ACCEPTANCE CRITERIA
* [x] Indoor locations should be available on the map to be searched and selected
* [x] The first location that the user clicks or searches for will become the destination point.
* [x] If location services are available, selecting the user's current location as the starting point is done automatically.
* [x] If the starting point isn't set, and the user places focus on the top text box, the next location that the user clicks or searches for will become the starting point.
* [x] After the starting and destination points are set, the user is allowed to start the travel plan
* [x] Once the user has started a travel plan, markers indicating the starting and destination location will be shown on the map.
* [x] Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=32%3A347)
* [x] Respect the [Definition of Done](https://github.com/RGPosadas/WayFinder/wiki/Definition-of-Done)


### USER FLOW ACCEPTANCE TEST
1. [x] Tap on the search bar
2. [x] User can search or tap on a point of interest to set it as the destination
3. [x] If location services are on, `starting location` is filled with the user's current location
   1. [x] If location services are not on, `starting location` is filled in by the user. First tap on the top text box and then either search for a POI or tap on a POI on the map
4. [x] Once the start and end locations are set, press on the now visible start travel button on the bottom right
5. [x] Markers for the start and end location will be shown on the map

No regressions encountered (there might be regression that I didn't manage to find). All tests passed. Well-documented. Much cleaner code than in the beginning. Good job!

RGPosadas reviewed (dismissed):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All tests pass
- [x] Implementation matches the corresponding mockup from Figma

### ACCEPTANCE CRITERIA
* [x] Indoor locations should be available on the map to be searched and selected
* [x] The first location that the user clicks or searches for will become the destination point.
* [x] If location services are available, selecting the user's current location as the starting point is done automatically.
* [x] If the starting point isn't set, and the user places focus on the top text box, the next location that the user clicks or searches for will become the starting point.
* [x] After the starting and destination points are set, the user is allowed to start the travel plan
* [x] Once the user has started a travel plan, markers indicating the starting and destination location will be shown on the map.
* [x] Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=32%3A347)
* [x] Respect the [Definition of Done](https://github.com/RGPosadas/WayFinder/wiki/Definition-of-Done)


### USER FLOW ACCEPTANCE TEST
1. [x] Tap on the search bar
2. [x] User can search or tap on a point of interest to set it as the destination
3. [x] If location services are on, `starting location` is filled with the user's current location
3. [x] If location services are not on, `starting location` is filled in by the user. First tap on the top text box and then either search for a POI or tap on a POI on the map
4. [x] Once the start and end locations are set, press on the now visible start travel button on the bottom right
5. [x] Markers for the start and end location will be shown on the map

### Notes
All blockers resolved! Nice work :)
Don't forget to rebase, other than that LGTM!

bit172 reviewed (dismissed):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All tests pass
- [x] Implementation matches the corresponding mockup from Figma

### ACCEPTANCE CRITERIA
* [x] Indoor locations should be available on the map to be searched and selected
* [x] The first location that the user clicks or searches for will become the destination point.
* [x] If location services are available, selecting the user's current location as the starting point is done automatically.
* [x] If the starting point isn't set, and the user places focus on the top text box, the next location that the user clicks or searches for will become the starting point.
* [x] After the starting and destination points are set, the user is allowed to start the travel plan
* [x] Once the user has started a travel plan, markers indicating the starting and destination location will be shown on the map.
* [x] Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=32%3A347)
* [x] Respect the [Definition of Done](https://github.com/RGPosadas/WayFinder/wiki/Definition-of-Done)

### USER FLOW ACCEPTANCE TEST
1. [x] Tap on the search bar
2. [x] User can search or tap on a point of interest to set it as the destination
3. [x] If location services are on, `starting location` is filled with the user's current location
   1. [x] If location services are not on, `starting location` is filled in by the user. First tap on the top text box and then either search for a POI or tap on a POI on the map
4. [x] Once the start and end locations are set, press on the now visible start travel button on the bottom right
5. [x] Markers for the start and end location will be shown on the map

r-saba committed:
Change building message position #18

r-saba committed:
Revert tests to js #18

r-saba committed:
Remove dependency #18

cristian-aldea committed:
Fix dependencies and bugs #18

cristian-aldea reviewed (approved):
Looks good to me! The new pushed invalidated my review, but my acceptance tests still apply.
